### PR TITLE
Add UsersFolder for media assets

### PR DESCRIPTION
**Pull request by felixhoi**
Fixes https://github.com/OrchardCMS/OrchardCore/issues/7576

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/OrchardCMS/OrchardCore?pullRequest=9357) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/OrchardCMS/OrchardCore?pullRequest=9357) before we can accept your contribution.<br/><br/>:x: felixhoi [sign now](https://cla.dotnetfoundation.org/OrchardCMS/OrchardCore?pullRequest=9357)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/OrchardCMS/OrchardCore?pullRequest=9357) it.</sub>

**Comment by agriffard** [comment]
Please fix the conflict.

**Comment by sebastienros** [comment]
Can you remove the Roles folder part, we should implement it this way instead: https://github.com/OrchardCMS/OrchardCore/issues/9369

File: src/OrchardCore.Modules/OrchardCore.Media/Controllers/AdminController.cs [hunk 1]
```diff
@@ -76,6 +77,19 @@ public async Task<ActionResult<IAsyncEnumerable<IFileStoreEntry>>> GetFolders(st
...
 
+            // create default folders if not exist
+
+            if (await _mediaFileStore.GetDirectoryInfoAsync(_mediaOptions.AssetsUsersFolder) == null)
+            {
+                await _mediaFileStore.TryCreateDirectoryAsync(_mediaOptions.AssetsUsersFolder);
+            }
+
+            if (await _authorizationService.AuthorizeAsync(User, Permissions.ManageOwnMedia)
+                && await _mediaFileStore.GetDirectoryInfoAsync($"{_mediaOptions.AssetsUsersFolder}/{User.Identity.Name}") == null)
```
[the remark this answers is not in the document]
Comment (deanmarcussen) [reply src/OrchardCore.Modules/OrchardCore.Media/Controllers/AdminController.cs:88]: The UserId by default is probably better, because there may be many scenarios where the you don't want the user name exposed (I wouldn't want my name as part of a url, for example).

However, other sites may wish to, or use a custom claim for the name, so I would suggest we could add a service, `IUserAssetFolderNameProvider` which takes a `ClaimsPrincipal`, and by default returns the `NameIdentifier` value.

In this way someone can replace the service if they wish to use a different naming convention.

File: src/OrchardCore.Modules/OrchardCore.Media/Services/AttachedMediaFieldsFolderAuthorizationHandler.cs [hunk 2]
```diff
@@ -11,21 +14,24 @@ namespace OrchardCore.Media.Services
     /// Check if the path passed as resource is inside the AttachedMediaFieldsFolder
     /// and in case it is, It checks if the user has ManageAttachedMediaFieldsFolder permission
     /// </summary>
-    public class AttachedMediaFieldsFolderAuthorizationHandler : AuthorizationHandler<PermissionRequirement>
+    public class AttachedMediaFieldsFolderAuthorizationHandler : AuthorizationHandler<PermissionRequirement>
```
[the remark this answers is not in the document]
Comment (felixhoi) [reply src/OrchardCore.Modules/OrchardCore.Media/Services/AttachedMediaFieldsFolderAuthorizationHandler.cs:17]: I am sorry I don't understand the purpose of original AttachedMediaFieldsFolderAuthorizationHandler. it seems to check the resource if under the path "media/mediafields". but under what circumstances will folder "media/mediafields" be created ? Can you explain more this ? Thank you.

[the remark this answers is not in the document]
Comment (deanmarcussen) [reply src/OrchardCore.Modules/OrchardCore.Media/Services/AttachedMediaFieldsFolderAuthorizationHandler.cs:17]: It's solely related to the Attached Media Field editor option on the media field (go to settings for a Media Field and select Attached as the editor).

When you use it, files are uploaded and selected, on an individual basis for the content item, rather than having the entire media library available.

So this permission handler is to allow management of the entire folder structure that gets created under it.

Does that make sense?

File: src/OrchardCore.Modules/OrchardCore.Media/Permissions.cs [hunk 3]
```diff
@@ -8,6 +8,9 @@ namespace OrchardCore.Media
     public class Permissions : IPermissionProvider
     {
         public static readonly Permission ManageMedia = new Permission("ManageMediaContent", "Manage Media");
+        public static readonly Permission ManageRootFolderMedia = new Permission("ManageRootFolderMediaContent", "Manage Media For Root Folder");
```
[the remark this answers is not in the document]
Comment (felixhoi) [reply src/OrchardCore.Modules/OrchardCore.Media/Permissions.cs]: I have removed it.

File: src/OrchardCore.Modules/OrchardCore.Media/Controllers/AdminController.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (felixhoi) [reply src/OrchardCore.Modules/OrchardCore.Media/Controllers/AdminController.cs]: IUserAssetFolderNameProvider added and by default return NameIdentifier.

File: src/OrchardCore.Modules/OrchardCore.Media/Services/AttachedMediaFieldsFolderAuthorizationHandler.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (felixhoi) [reply src/OrchardCore.Modules/OrchardCore.Media/Services/AttachedMediaFieldsFolderAuthorizationHandler.cs]: Thank you. I have refactored AttachedMediaFieldsFolderAuthorizationHandler and rename to ManageMediaFolderAuthorizationHandler, let me know if any improvement need. thank you.

File: src/OrchardCore.Modules/OrchardCore.Media/Views/Admin/Options.cshtml [hunk 4]
```diff
@@ -72,4 +72,9 @@
         <code><input asp-for="AssetsPath" type="text" class="form-control-plaintext" /></code>
         <span class="hint">@T["The path used to store media assets. The path can be relative to the tenant's App_Data folder, or absolute."]</span>
     </div>
+    <label asp-for="AssetsUsersFolder" class="col-sm-2 col-form-label">@T["Assets users folder"]</label>
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Media/Views/Admin/Options.cshtml:75]: I would not allow to change it through UI. Only from app settings.

Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.Media/Views/Admin/Options.cshtml:75]: Or not at all, like we can't change `media`

Comment (felixhoi) [reply src/OrchardCore.Modules/OrchardCore.Media/Views/Admin/Options.cshtml:75]: https://docs.orchardcore.net/en/latest/docs/reference/modules/Media/#configuration
media path can be change through appsettings.json, the UI only for display settings, can't save changes.

File: src/OrchardCore.Modules/OrchardCore.Media/Permissions.cs [hunk 5]
```diff
@@ -8,15 +8,20 @@ namespace OrchardCore.Media
     public class Permissions : IPermissionProvider
     {
         public static readonly Permission ManageMedia = new Permission("ManageMediaContent", "Manage Media");
-        public static readonly Permission ManageAttachedMediaFieldsFolder = new Permission("ManageAttachedMediaFieldsFolder", "Manage Attached Media Fields Folder");
+        public static readonly Permission ManageMediaFolder = new Permission("ManageMediaFolder", "Manage All Media Folder");
+        public static readonly Permission ManageOthersMedia = new Permission("ManageOthersMediaContent", "Manage Media For Others", new[] { ManageMediaFolder });
+        public static readonly Permission ManageOwnMedia = new Permission("ManageOwnMediaContent", "Manage Own Media" , new[] { ManageOthersMedia });        
+        public static readonly Permission ManageAttachedMediaFieldsFolder = new Permission("ManageAttachedMediaFieldsFolder", "Manage Attached Media Fields Folder", new[] { ManageMediaFolder});
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Media/Permissions.cs:14]: Looking good, but I have some questions around the new permission structure

By this new permission structure `ManageMediaFolder` is now the key permission required to view general media in the media folder (i.e. media that is not Own or Others?)

Is that correct, or have I misread?

If so, `ManageAttachedMediaFieldsFolder` should not be implied by it, as that is a specific admin permission.
Same with `ManageOthersMedia`, should be specific permission.

An Editor and the other users should have permission to view standard files in the media library, by default.

Otherwise the behaviour is going to change significantly for current users.

The key for this pr, is for existing behaviour to remain the same, and to allow consumers to reduce permission, if they wish to restrict users to just their OwnMedia, and make use of the feature.

Comment (felixhoi) [reply src/OrchardCore.Modules/OrchardCore.Media/Permissions.cs:14]: `ManageMediaFolder` is like a super admin permission, can access all the folder under media, include own,  others folder, attachedMediaFields folder. 

The key permission is still `ManageMedia` , Not 'ManageMediaFolder'.
at `ManageMediaFolderAuthorizationHandler.cs`, firstly the permission is set to Permissions.ManageMedia, then the permission morphed according the folder name.

if not folder name(ie. attachedMediaField, own, others) match, that means user request a general media resource and the request permission will be `ManageMedia`. So it will not effect current users, current users should have `ManageMedia` 

let me know if it still not easy to understand, or anything need to improve. Thank you.

File: src/OrchardCore.Modules/OrchardCore.Media/Permissions.cs [hunk 6]
```diff
@@ -8,15 +8,20 @@ namespace OrchardCore.Media
     public class Permissions : IPermissionProvider
     {
         public static readonly Permission ManageMedia = new Permission("ManageMediaContent", "Manage Media");
-        public static readonly Permission ManageAttachedMediaFieldsFolder = new Permission("ManageAttachedMediaFieldsFolder", "Manage Attached Media Fields Folder");
+        public static readonly Permission ManageMediaFolder = new Permission("ManageMediaFolder", "Manage All Media Folder");
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Media/Permissions.cs:11]:
```suggestion
        public static readonly Permission ManageMediaFolder = new Permission("ManageMediaFolder", "Manage All Media Folders");
```

File: src/OrchardCore.Modules/OrchardCore.Media/Permissions.cs (hunk 5 above)
Comment (deanmarcussen) [reply src/OrchardCore.Modules/OrchardCore.Media/Permissions.cs:14]: ok, so that makes sense, but it wasn't obvious.

File: src/OrchardCore.Modules/OrchardCore.Media/Permissions.cs [hunk 7]
```diff
@@ -31,12 +36,19 @@ public IEnumerable<PermissionStereotype> GetDefaultStereotypes()
...
                     Name = "Administrator",
-                    Permissions = new[] { ManageMedia, ManageAttachedMediaFieldsFolder, ManageMediaProfiles, ViewMediaOptions }
+                    Permissions = new[] {
+                        ManageMedia,
+                        ManageMediaFolder,
+                        ManageOthersMedia,
+                        ManageOwnMedia,
+                        ManageAttachedMediaFieldsFolder,
+                        ManageMediaProfiles,
+                        ViewMediaOptions }
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Media/Permissions.cs:46]: Here, the permissions added to the `Administrator` role by default, should only include the `ManageMediaFolder`, rather than the `ManageMedia`, `ManageOthersMedia`, and other related implied permissions.

This is because many of these permissions, are implied by something else, so when we add them as defaults, we should add the most senior permission.

`ManagedMediaProfiles`, and `ViewMediaOptions` look like they should remain there, as they are not implied by anything.

Comment (felixhoi) [reply src/OrchardCore.Modules/OrchardCore.Media/Permissions.cs:42]: ok, deleted.

File: src/OrchardCore.Cms.Web/OrchardCore.Cms.Web.csproj [hunk 8]
```diff
@@ -11,6 +11,7 @@
     <WarningsNotAsErrors>612,618</WarningsNotAsErrors>
     <AspNetCoreHostingModel>InProcess</AspNetCoreHostingModel>
     <PreserveCompilationReferences>true</PreserveCompilationReferences>
+    <UserSecretsId>3d0ddb2b-6694-44de-99b9-2085feec2408</UserSecretsId>
```
Comment (deanmarcussen) [inline src/OrchardCore.Cms.Web/OrchardCore.Cms.Web.csproj:14]: Can you revert this change?

File: src/OrchardCore.Cms.Web/.config/dotnet-tools.json [hunk 9]
```diff
@@ -0,0 +1,12 @@
+{
+  "version": 1,
+  "isRoot": true,
+  "tools": {
+    "dotnet-ef": {
+      "version": "5.0.6",
+      "commands": [
+        "dotnet-ef"
```
Comment (deanmarcussen) [inline src/OrchardCore.Cms.Web/.config/dotnet-tools.json:8]: Can you revert this change?

File: src/OrchardCore.Modules/OrchardCore.Media/Services/IUserAssetFolderNameProvider.cs [hunk 10]
```diff
@@ -0,0 +1,14 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Claims;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace OrchardCore.Media.Services
+{
+    public interface IUserAssetFolderNameProvider
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Media/Services/IUserAssetFolderNameProvider.cs:10]: Can you move this interface to `OrchardCore.Media.Abstractions` (in the root folder, with `OrchardCore.Media` as the namespace please.

**Review by deanmarcussen — COMMENTED** [review]
Looking good thanks @felixhoi just a couple of things to tidy up

File: src/OrchardCore.Cms.Web/.config/dotnet-tools.json (hunk 9 above)
Comment (felixhoi) [reply src/OrchardCore.Cms.Web/.config/dotnet-tools.json]: ok

File: src/OrchardCore.Cms.Web/OrchardCore.Cms.Web.csproj (hunk 8 above)
Comment (felixhoi) [reply src/OrchardCore.Cms.Web/OrchardCore.Cms.Web.csproj]: ok

File: src/OrchardCore.Modules/OrchardCore.Media/Services/IUserAssetFolderNameProvider.cs (hunk 10 above)
Comment (felixhoi) [reply src/OrchardCore.Modules/OrchardCore.Media/Services/IUserAssetFolderNameProvider.cs]: done

File: src/OrchardCore.Cms.Web/OrchardCore.Cms.Web.csproj [hunk 11]
```diff
@@ -29,4 +29,4 @@
 
   <Import Project="..\OrchardCore.Build\OrchardCore.Commons.targets" />
 
-</Project>
+</Project>
```
Comment (deanmarcussen) [inline src/OrchardCore.Cms.Web/OrchardCore.Cms.Web.csproj:32]: Please revert the uneeded changes to this file

Comment (felixhoi) [reply src/OrchardCore.Cms.Web/OrchardCore.Cms.Web.csproj]: done

**Review by deanmarcussen — APPROVED** [review]
Thank you
